### PR TITLE
#271 Adding value-time-lock support to the batch writer and observer

### PR DESCRIPTION
**Pull request by mudiali**

### CODE REVIEW

File: lib/core/versions/latest/BatchWriter.ts [hunk 1]
```diff
@@ -77,4 +83,10 @@ export default class BatchWriter implements IBatchWriter {
     // Remove written operations from queue after batch writing has completed successfully.
     await this.operationQueue.dequeue(queuedOperations.length);
   }
+
+  private getNumberOfOpsToWrite (valueTimeLock: ValueTimeLockModel | undefined, normalizedFee: number): number {
+    const numberOfOpsAllowed = ValueTimeLockVerifier.calculateMaxNumberOfOpsAllowed(valueTimeLock, normalizedFee);
+
+    return Math.min(numberOfOpsAllowed, ProtocolParameters.maxOperationsPerBatch);
```
Comment (isaacJChen) [inline lib/core/versions/latest/BatchWriter.ts]: Should we log some sort of message here if the maxOperationPerBatch is lower than numberOfOpsAllowed so the admin knows when they are "wasting" money in lock?

File: lib/core/Blockchain.ts [hunk 2]
```diff
@@ -218,11 +218,6 @@ export default class Blockchain implements IBlockchain {
     const responseBody = await JsonAsync.parse(responseBodyString);
 
     if (response.status === HttpStatus.NOT_FOUND) {
-
-      if (responseBody.code === SharedErrorCode.ValueTimeLockInPendingState) {
-        throw new SidetreeError(SharedErrorCode.ValueTimeLockInPendingState);
```
Comment (thehenrytsai) [inline lib/core/Blockchain.ts]: Move `ValueTimeLockInPendingState` back to bicoin service's own `ErrorCode`.

File: lib/core/versions/latest/AnchorFile.ts [hunk 3]
```diff
@@ -61,6 +63,13 @@ export default class AnchorFile {
       throw new SidetreeError(ErrorCode.AnchorFileMissingOperationsProperty);
     }
 
+    // This property is optional so if there's an additional propertye besides
+    // the required ones then it must be this optional one.
+    if (anchorFileProperties.length === maxAllowedPropertyCount &&
+       !anchorFileModel.hasOwnProperty('writerLock')) {
+      throw new SidetreeError(ErrorCode.AnchorFileHasUnknownProperty);
```
Comment (thehenrytsai) [inline lib/core/versions/latest/AnchorFile.ts]: Nit: In principal we try to never reuse an error code, so consider either:

1. Create a new error code. or
1. Attach custom error message, or
1. Follow the pattern found in other validation handling optional parameters, e.g.
```javascript
    const allowedProperties = new Set(['batchFileHash', 'updateOperations']);
    for (let property in mapFile) {
      if (!allowedProperties.has(property)) {
        throw new SidetreeError(ErrorCode.MapFileHasUnknownProperty);
      }
    }
```

File: lib/core/versions/latest/AnchorFile.ts [hunk 4]
```diff
@@ -170,6 +180,7 @@ export default class AnchorFile {
     });
 
     const anchorFileModel = {
+      writerLock: writerLock,
```
Comment (thehenrytsai) [inline lib/core/versions/latest/AnchorFile.ts]: Nit:
```suggestion
      writerLock,
```

File: lib/core/versions/latest/ValueTimeLockVerifier.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+   * Calculates the maximum number of operations allowed to be written for the given lock information. If
+   * there is no lock then it returns the number of operations which do not require a lock.
+   *
+   * @param valueTimeLock The lock object if exists
+   * @param normalizedFee The normalized fee for the current block
+   */
+  public static calculateMaxNumberOfOpsAllowed (valueTimeLock: ValueTimeLockModel | undefined, normalizedFee: number) {
+
+    if (valueTimeLock === undefined) {
+      return ProtocolParameters.maxNumberOfOpsForNoValueTimeLock;
```
Comment (thehenrytsai) [inline lib/core/versions/latest/ValueTimeLockVerifier.ts]: In the spirit of not introducing further parameters unless we must, do we want to derive `maxNumberOfOpsForNoValueTimeLock` from `normalizedFeeToPerOperationFeeMultiplier`?

If we change `normalizedFeeToPerOperationFeeMultiplier` to `normalizedFeeToPerOperationFeeDivisor`, then we can just use the value as is.

File: lib/core/versions/latest/ValueTimeLockVerifier.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    // Using the following formula:
+    //  requiredLockAmount = normalizedfee * normalizedFeeMultipier * numberOfOps * valueTimeLockMultiplier
+    //
+    // We are going to find the numberOfOps given the requiredLockAmount
+    const feePerOperation = normalizedFee * ProtocolParameters.normalizedFeeToPerOperationFeeMultiplier;
+    const numberOfOpsAllowed = valueTimeLock.amountLocked / (feePerOperation * ProtocolParameters.valueTimeLockAmountMultiplier);
```
Comment (thehenrytsai) [inline lib/core/versions/latest/ValueTimeLockVerifier.ts]: We should return the max free ops count if the `numberOfOpsAllowed` calculated turns out to be lower.

File: lib/core/versions/latest/BatchWriter.ts [hunk 7]
```diff
@@ -77,4 +83,10 @@ export default class BatchWriter implements IBatchWriter {
     // Remove written operations from queue after batch writing has completed successfully.
     await this.operationQueue.dequeue(queuedOperations.length);
   }
+
+  private getNumberOfOpsToWrite (valueTimeLock: ValueTimeLockModel | undefined, normalizedFee: number): number {
+    const numberOfOpsAllowed = ValueTimeLockVerifier.calculateMaxNumberOfOpsAllowed(valueTimeLock, normalizedFee);
```
Comment (thehenrytsai) [inline lib/core/versions/latest/BatchWriter.ts]: Consider just let `calculateMaxNumberOfOpsAllowed` perform the `Math.min()` check.

File: lib/core/versions/latest/TransactionProcessor.ts [hunk 8]
```diff
@@ -88,6 +90,18 @@ export default class TransactionProcessor implements ITransactionProcessor {
...
     }
 
+    // Verify required lock if one was needed.
+    const valueTimeLock = anchorFile.model.writerLock
+                          ? await this.blockchain.getValueTimeLock(anchorFile.model.writerLock)
+                          : undefined;
+
+    ValueTimeLockVerifier.verifyLockAmountAndThrowOnError(
+      valueTimeLock,
+      operationCountInAnchorFile,
```
Comment (thehenrytsai) [inline lib/core/versions/latest/TransactionProcessor.ts]: Sanity: should this be `paidOperationCount`?

File: lib/core/versions/latest/ValueTimeLockVerifier.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+   * @param valueTimeLock The value time lock object used for verificiation.
+   * @param numberOfOperations The target number of operations.
+   * @param normalizedFee The normalized fee for the target block.
+   * @param targetTransactionTime The transaction time where the operations were written.
+   */
+  public static verifyLockAmountAndThrowOnError (
+    valueTimeLock: ValueTimeLockModel | undefined,
+    numberOfOperations: number,
+    normalizedFee: number,
+    targetTransactionTime: number,
```
Comment (thehenrytsai) [inline lib/core/versions/latest/ValueTimeLockVerifier.ts]: Consider rename:
```suggestion
    sidetreeTransactionTime: number,

```

File: lib/core/versions/latest/ValueTimeLockVerifier.ts [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    targetTransactionWriter: string): void {
```
Comment (thehenrytsai) [inline lib/core/versions/latest/ValueTimeLockVerifier.ts]: Consider rename:
```suggestion
    sidetreeTransactionWriter: string): void {

```

File: lib/core/versions/latest/ValueTimeLockVerifier.ts [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+          targetTransactionTime >= valueTimeLock.unlockTransactionTime) {
+        throw new SidetreeError(
+          ErrorCode.ValueTimeLockTargetTransactionTimeOutsideLockRange,
+          `Target block: ${targetTransactionTime}; lock start time: ${valueTimeLock.lockTransactionTime}; unlock time: ${valueTimeLock.unlockTransactionTime}`);
+      }
+    }
+
+    const maxNumberOfOpsAllowed = this.calculateMaxNumberOfOpsAllowed(valueTimeLock, normalizedFee);
+
+    if (numberOfOperations > maxNumberOfOpsAllowed) {
```
Comment (thehenrytsai) [inline lib/core/versions/latest/ValueTimeLockVerifier.ts]: Sanity: doesn't look right, 'coz this means a spammer can bypass the protocol maximum if we don't check for it. Come back to the comment earlier about considering having the protocol maximum check within `calculateMaxNumberOfOpsAllowed()`.

File: lib/core/versions/latest/models/ProtocolParameters.ts [hunk 12]
```diff
@@ -14,6 +14,8 @@ export default interface ProtocolParameters {
   maxEncodedHashStringLength: number;
   /** Max number of operations observed per transaction time */
   maxNumberOfOpsPerTransactionTime: number;
+  /** Maximum number of operations allowed with no lock. */
+  maxNumberOfOpsForNoValueTimeLock: number;
```
Comment (thehenrytsai) [inline lib/core/versions/latest/models/ProtocolParameters.ts]: Prefer naming consistency: (not sure how`maxNumberOfOpsPerTransactionTime` fell through review cracks). Open to have the change `operations` to `ops` throughout discussion 'coz I get it gets rather wordy....
```suggestion
  maxNumberOfOperationsForNoValueTimeLock: number;
```

File: lib/core/versions/latest/ValueTimeLockVerifier.ts [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+export default class ValueTimeLockVerifier {
+
+  /**
+   * Calculates the maximum number of operations allowed to be written for the given lock information. If
+   * there is no lock then it returns the number of operations which do not require a lock.
+   *
+   * @param valueTimeLock The lock object if exists
+   * @param normalizedFee The normalized fee for the current block
+   */
+  public static calculateMaxNumberOfOpsAllowed (valueTimeLock: ValueTimeLockModel | undefined, normalizedFee: number) {
```
Comment (thehenrytsai) [inline lib/core/versions/latest/ValueTimeLockVerifier.ts]: Prefer consistency with existing code. Open to have the change from `operations` to `ops` throughout discussion 'coz I get it gets rather wordy....
```suggestion
  public static calculateMaxNumberOfOperationsAllowed (valueTimeLock: ValueTimeLockModel | undefined, normalizedFee: number) {

```

File: lib/core/versions/latest/ValueTimeLockVerifier.ts [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+          ErrorCode.ValueTimeLockTransactionWriterLockOwnerMismatch,
+          `Transaction writer: ${targetTransactionWriter} - Lock owner: ${valueTimeLock.owner}`);
+      }
+
+      // Check the lock duration
+      if (targetTransactionTime < valueTimeLock.lockTransactionTime ||
+          targetTransactionTime >= valueTimeLock.unlockTransactionTime) {
+        throw new SidetreeError(
+          ErrorCode.ValueTimeLockTargetTransactionTimeOutsideLockRange,
+          `Target block: ${targetTransactionTime}; lock start time: ${valueTimeLock.lockTransactionTime}; unlock time: ${valueTimeLock.unlockTransactionTime}`);
```
Comment (thehenrytsai) [inline lib/core/versions/latest/ValueTimeLockVerifier.ts]:
```suggestion
          `Sidetree transaction time: ${targetTransactionTime}; lock start time: ${valueTimeLock.lockTransactionTime}; unlock time: ${valueTimeLock.unlockTransactionTime}`);

```

File: lib/core/versions/latest/ValueTimeLockVerifier.ts [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+   * @param targetTransactionTime The transaction time where the operations were written.
+   */
+  public static verifyLockAmountAndThrowOnError (
+    valueTimeLock: ValueTimeLockModel | undefined,
+    numberOfOperations: number,
+    normalizedFee: number,
+    targetTransactionTime: number,
+    targetTransactionWriter: string): void {
+
+    if (valueTimeLock) {
```
Comment (thehenrytsai) [inline lib/core/versions/latest/ValueTimeLockVerifier.ts]: Discussion: do we want to let the verification succeed if the lock verification fails but the actual operation count is within the "free" limit?

File: lib/core/versions/latest/BatchWriter.ts (hunk 1 above)
Comment (mudiali) [reply lib/core/versions/latest/BatchWriter.ts]: Adding a log statement here.

File: lib/core/versions/latest/AnchorFile.ts (hunk 3 above)
Comment (mudiali) [reply lib/core/versions/latest/AnchorFile.ts]: Really good point. Will update.

File: lib/core/versions/latest/AnchorFile.ts (hunk 4 above)
Comment (mudiali) [reply lib/core/versions/latest/AnchorFile.ts]: Renaming it to `writerLockId` to match the `.getWriterLock(...)` function that is in the `IBlockchain`.

File: lib/core/versions/latest/BatchWriter.ts (hunk 7 above)
Comment (mudiali) [reply lib/core/versions/latest/BatchWriter.ts]: I want to keep the ValueTimeLockVerifier to only care about the lock related checks. The `maxOperationsPerBatch` is a protocol check which belongs outside of the verifier IMO. I am refactoring the code to make it more clearer.

File: lib/core/versions/latest/ValueTimeLockVerifier.ts (hunk 11 above)
Comment (mudiali) [reply lib/core/versions/latest/ValueTimeLockVerifier.ts]: The `maxOperationPerBatch` check is performed outside in `TransactionProcessor.downloadAndVerifyAnchorFile(...)` function.

File: lib/core/versions/latest/models/ProtocolParameters.ts (hunk 12 above)
Comment (mudiali) [reply lib/core/versions/latest/models/ProtocolParameters.ts]: Renaming.

File: lib/core/versions/latest/ValueTimeLockVerifier.ts (hunk 13 above)
Comment (mudiali) [reply lib/core/versions/latest/ValueTimeLockVerifier.ts]: Changing the name.

File: lib/core/versions/latest/TransactionProcessor.ts (hunk 8 above)
Comment (mudiali) [reply lib/core/versions/latest/TransactionProcessor.ts]: Good catch. Done.

File: lib/core/versions/latest/ValueTimeLockVerifier.ts (hunk 5 above)
Comment (mudiali) [reply lib/core/versions/latest/ValueTimeLockVerifier.ts]: Keeping the two values separate to keep the max-free-ops and normalized-fee separate.

File: lib/core/versions/latest/ValueTimeLockVerifier.ts (hunk 15 above)
Comment (mudiali) [reply lib/core/versions/latest/ValueTimeLockVerifier.ts]: Good idea. Done.

Comment (mudiali) [reply lib/core/versions/latest/ValueTimeLockVerifier.ts]: Done.

File: lib/core/versions/latest/BatchWriter.ts [hunk 16]
```diff
@@ -85,8 +85,14 @@ export default class BatchWriter implements IBatchWriter {
   }
 
   private getNumberOfOpsToWrite (valueTimeLock: ValueTimeLockModel | undefined, normalizedFee: number): number {
```
Comment (thehenrytsai) [inline lib/core/versions/latest/BatchWriter.ts]: Nit:
```suggestion
  private getNumberOfOperationsToWrite (valueTimeLock: ValueTimeLockModel | undefined, normalizedFee: number): number {
```

**Pull request merged**